### PR TITLE
added three tests to cover changes made by the pull request #1361

### PR DESCRIPTION
**Pull request by GreenfishK**
# Fixes
SUBSTR: The 'length' parameter was not included. Now it is
SUBSTR: Did not work with group variables. E.g. when ?x was used in group by and in substr, ?x would appear as __aggr_1 in Builtin_substr in the query algebra tree. Group variables in built in functions are now replaced properly
CONCAT: Nested expressions with group variables in CONCAT are now translated properly
function header: The parameter query_algebra is not optional anymore

# Changes
algebra.py: copied the code from algebra in PR #1361 and made another correction in the Builtin_SUBSTR section so that nested arguments are resolved correctly.
test_functions__functions_on_srings.txt: one line was added in the select clause to test the substr function with the position and length parameter. 
test_graph_patterns__group_and_substr.txt: Added a group variable (from the group by clause) in the substr function of the select clause to test whether such queries get properly translated.
test_graph_patterns__group_and_nested_concat.txt: added a nested concat expression with a group variable. This test fails with a TypeError.

  -
  -
  -

### CODE REVIEW

**Comment by GreenfishK** [comment]
@aucampia  @nicholascar  How should we proceed here? Should I make the bugfix in this PR or make a separate pull request for the bug fix?

**Comment by aucampia** [comment]
> How should we proceed here? Should I make the bugfix in this PR or make a separate pull request for the bug fix?

@GreenfishK I think it is best to combine this with the fix if you have it, which as far as I understand in this case you do.

Sometimes we will merge failing tests, but then they must be marked with xfail ([e.g. here](https://github.com/RDFLib/rdflib/blob/1226b52da1b839d865f4ae196dd32c1d374f9a2c/test/test_translate_algebra.py#L183-L189)) so that they don't block the CI system, but generally this would only be if we don't have a fix at hand. xfail test make it easier to illustrate the problem to someone else and then later on do a handover, and it is also good for new contributors to get into things as they have less to learn. Sufficently annotated xfail markers (i.e. ones with `raises`) also make it possible to see if something breaks further (i.e. if a different exception is raised), and it also makes it possible to see if something was inadvertently fixed as they still run and if they do pass they get reported as xpass (unexpected pass)

**Comment by aucampia** [comment]
Change looks good though, but to summarise, it would be best to combine it with #1361 by either adding it to that PR (probably simplest), or adding the changes from that PR to this and updating the description here.

**Comment by GreenfishK** [comment]
@aucampia There seem to be still some issues with the translateAlgebra function and I am going nuts while trying to figure out how to activate the logger in order to debug it. I could not find any pyproject.toml or pytest.ini file where I could set the level to DEBUG. I also tried running pytest via CLI and with some log options activated but it did not work. Do you have any suggestions?
Thanks!

**Comment by GreenfishK** [comment]
> @aucampia There seem to be still some issues with the translateAlgebra function and I am going nuts while trying to figure out how to activate the logger in order to debug it. I could not find any pyproject.toml or pytest.ini file where I could set the level to DEBUG. I also tried running pytest via CLI and with some log options activated but it did not work. Do you have any suggestions? Thanks!

Oh, nevermind. I figured it out. I just put in tox.ini. Thanks anyway

**Comment by aucampia** [comment]
depends on how you run it, I run on my computer like this:

```bash
## setup venv
\rm -rv .venv
python3.8 -m venv .venv
.venv/bin/python3 -m pip install -r requirements.txt -r requirements.dev.txt -r requirements.dev-extra.txt

## run test with debug logging (--log-level DEBUG) and report for all tests (-rA)
.venv/bin/python3 -m pytest --log-level DEBUG -rA
```

But yes if you change tox.ini:

```diff
diff --git a/tox.ini b/tox.ini
index c09483da..38b534b0 100644
--- a/tox.ini
+++ b/tox.ini
@@ -9,7 +9,7 @@ commands =
     {envpython} -m mypy rdflib --show-error-context --show-error-codes
     {envpython} setup.py clean --all
     {envpython} setup.py build
-    {envpython} -m pytest
+    {envpython} -m pytest --log-level DEBUG -rA
 deps =
        -rrequirements.txt
        -rrequirements.dev.txt
```

Then it will do the same, and you can do the same by editing setup.cfg:

```diff
diff --git a/setup.cfg b/setup.cfg
index 3e59cdf2..5990ade2 100644
--- a/setup.cfg
+++ b/setup.cfg
@@ -39,9 +39,13 @@ addopts =
    --ignore=admin
    --ignore=rdflib/extras/external_graph_libs.py
    --ignore-glob=docs/*.py
+   -rA
 doctest_optionflags = ALLOW_UNICODE
 filterwarnings =
     # The below warning is a consequence of how pytest doctest detects mocks and how DefinedNamespace behaves when an undefined attribute is being accessed.
     ignore:Code. pytest_mock_example_attribute_that_shouldnt_exist is not defined in namespace .*:UserWarning
     # The below warning is a consequence of how pytest detects fixtures and how DefinedNamespace behaves when an undefined attribute is being accessed.
     ignore:Code. _pytestfixturefunction is not defined in namespace .*:UserWarning
+log_cli = true
+log_cli_level = "DEBUG"
```

or run with:

```
PYTEST_ADDOPTS="-rA --log-level DEBUG" tox -e py37
```

I will add it to developers.rst when I have time.

**Comment by GreenfishK** [comment]
I just added this to tox.ini
```
[pytest]
log_cli = 1
log_cli_level = DEBUG 
log_cli_format = %(asctime)s [%(levelname)8s] %(message)s (%(filename)s:%(lineno)s) 
log_cli_date_format=%Y-%m-%d %H:%M:%S 
```
I am running it comfortably using PyCharm

**Comment by GreenfishK** [comment]
> By the way, I am on matrix/gitter if you want help, won't always respond immediately but I do have notifications turned on: https://gitter.im/RDFLib/rdflib / https://matrix.to/#/#RDFLib_rdflib:gitter.im

Thank you! For now I am good, I just solved the issues and will make another commit to this PR and add the description from #1361. This way it was somehow easier for me.

**Comment by aucampia** [comment]
I don't know what is wrong with drone, it looks like everything is fine though, maybe Nic can retrigger drone somehow, but you can also re-trigger it by amending your latest commit and doing a force push:

```bash
git commit --amend --date=now
git push --force
```

**Comment by GreenfishK** [comment]
> I don't know what is wrong with drone, it looks like everything is fine though, maybe Nic can retrigger drone somehow, but you can also re-trigger it by amending your latest commit and doing a force push:
> 
> ```shell
> git commit --amend --date=now
> git push --force
> ```

I tried it but it didn't work. hm...

**Comment by aucampia** [comment]
@nicholascar @ashleysommer any idea what is up with drone?

**Comment by aucampia** [comment]
@GreenfishK I'm fine if all other tests pass except the drone tests, so I would not let that hold you back if it is.

**Comment by aucampia** [comment]
Please let me know once this is ready for review, I don't want to review if you are still working on it and I'm assuming you are still working on it given you mentioned you add the description from #1361.

**Comment by GreenfishK** [comment]
> Please let me know once this is ready for review, I don't want to review if you are still working on it and I'm assuming you are still working on it given you mentioned you add the description from #1361.

Hi @aucampia, it is ready for review. I actually already added the description in the section "Fixes"

**Comment by aucampia** [comment]
@GreenfishK Apologies for not noticing, I will have a look tonight, thanks for the patience and effort.

**Comment by GreenfishK** [comment]
> @GreenfishK Apologies for not noticing, I will have a look tonight, thanks for the patience and effort.

Thank's for your support!

File: rdflib/plugins/sparql/algebra.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 """
 
 import functools
+import logging
```
Comment (aucampia) [inline rdflib/plugins/sparql/algebra.py:9]:
```suggestion
```

I think it is best to remove the logging.

File: rdflib/plugins/sparql/algebra.py [hunk 2]
```diff
@@ -847,9 +848,16 @@ def find_nth(haystack, needle, n):
...
             file.write(filedata)
 
+    aggr_vars = collections.defaultdict(list)  # type: dict
+
     def convert_node_arg(node_arg):
         if isinstance(node_arg, Identifier):
-            return node_arg.n3()
+            if node_arg in aggr_vars.keys():
+                grp_var = aggr_vars[node_arg].pop(0).n3()
+                logging.debug(grp_var)
```
Comment (aucampia) [inline rdflib/plugins/sparql/algebra.py:857]: best to remove the logging

```suggestion
```

File: rdflib/plugins/sparql/algebra.py [hunk 3]
```diff
@@ -1294,7 +1304,8 @@ def sparql_query_text(node):
                 )
             elif node.name.endswith("Builtin_CONCAT"):
                 expr = "CONCAT({vars})".format(
-                    vars=", ".join(elem.n3() for elem in node.arg)
+                    #vars=", ".join(elem.n3() for elem in node.arg)
```
Comment (aucampia) [inline rdflib/plugins/sparql/algebra.py:1307]:
```suggestion
```

Comment can be removed.

File: tox.ini [hunk 4]
```diff
@@ -45,3 +45,9 @@ setenv =
     PYTHONHASHSEED = 0
 commands =
     sphinx-build -n -T -b html -d {envtmpdir}/doctrees docs docs/_build/html
+
+[pytest]
+log_cli = 1
+log_cli_level = DEBUG
```
Comment (aucampia) [inline tox.ini:51]:
```suggestion
# log_cli = true
# log_cli_level = DEBUG
```

I would rather comment this out for now. While it is helpful during debugging it is not quite ideal in all cases and can be done via command line interface also. Leaving it as a comment will maybe make it easier for people to find if they don't know pytest that well.

Could also remove it, up to you, but just don't want it using debug and doing live logging always.

**Review by aucampia — COMMENTED** [review]
Change looks good, I made some small suggestions (relating to comments and logging) but otherwise I think it is okay to merge.

**Comment by GreenfishK** [comment]
> Change looks good, I made some small suggestions (relating to comments and logging) but otherwise I think it is okay to merge.

Ok, I took over all your suggestions and committed the changes.

**Review by aucampia — APPROVED** [review]
Looks good to me, thank you for the effort @GreenfishK.

**Review by nicholascar — APPROVED** [review]
Thank you both @GreenfishK and @aucampia for all your work on this PR!

**Pull request merged**